### PR TITLE
Init HttpListener Implement

### PR DESCRIPTION
**Pull request by younger15**

### CODE REVIEW

File: CMakeLists.txt
Comment (john0312) [inline CMakeLists.txt]: Please add license header.

File: CMakeLists.txt
Comment (john0312) [inline CMakeLists.txt]: Please remove redundant comment.

File: CMakeLists.txt
Comment (john0312) [inline CMakeLists.txt]: Consider using a separate PR for cmake.

File: CMakeLists.txt
Comment (john0312) [inline CMakeLists.txt]: Please add a new line at the end.

File: src/Implement/HttpListener.cpp
Comment (john0312) [inline src/Implement/HttpListener.cpp]: No need to duplicate function comment in .cpp.

File: src/Implement/HttpListener.cpp
Comment (john0312) [inline src/Implement/HttpListener.cpp:28]: Please use a logging library, or remove redundant code.

Comment (john0312) [reply src/Implement/HttpListener.cpp:28]: Or leave printf as is and add a TODO.

File: src/Implement/HttpListener.cpp
Comment (john0312) [inline src/Implement/HttpListener.cpp]: Consider doing recv in the child process.

File: src/Implement/HttpListener.cpp
Comment (john0312) [inline src/Implement/HttpListener.cpp]: Request might be longer than 2048 bytes.

File: src/Implement/HttpListener.cpp
Comment (john0312) [inline src/Implement/HttpListener.cpp]: Please remove redundant code.

File: src/Implement/UnitTestHttpListener.cpp
Comment (john0312) [inline src/Implement/UnitTestHttpListener.cpp]: Usually we put the mock in a separate header.

File: src/Implement/UnitTestHttpListener.cpp
Comment (john0312) [inline src/Implement/UnitTestHttpListener.cpp:52]: As discussed in class, we want to test HttpListener's functionality, not the mock.

File: CMakeLists.txt
Comment (younger15) [reply CMakeLists.txt]: Done

File: src/Implement/HttpListener.cpp
Comment (younger15) [reply src/Implement/HttpListener.cpp]: If listener only accept, than what does listener need to test? Only port number is under control. File descriptor is assigned automatically.

File: src/Implement/HttpListener.cpp
Comment (younger15) [reply src/Implement/HttpListener.cpp]: May be move to request handler with child process

File: src/Implement/HttpListener.cpp
Comment (younger15) [reply src/Implement/HttpListener.cpp]: Done

File: src/Implement/UnitTestHttpListener.cpp
Comment (younger15) [reply src/Implement/UnitTestHttpListener.cpp]: Removed

File: src/Implement/HttpListener.cpp
Comment (younger15) [reply src/Implement/HttpListener.cpp]: Done

File: CMakeLists.txt
Comment (younger15) [reply CMakeLists.txt]: CMakeList removed

File: CMakeLists.txt
Comment (younger15) [reply CMakeLists.txt]: CMakeList removed

Comment (younger15) [reply CMakeLists.txt]: CMakeList removed

File: CMakeLists.txt
Comment (younger15) [reply CMakeLists.txt]: CMakeList removed

File: src/Implement/HttpListener.cpp
Comment (john0312) [reply src/Implement/HttpListener.cpp:28]: It would be preferable to uncomment the printf so that you'll still get error message before the TODO is done.

File: src/Implement/UnitTestHttpListener.cpp
Comment (john0312) [reply src/Implement/UnitTestHttpListener.cpp:52]: This is still not testing the HttpListener's functionality.

You need to use a TCP Client to connect to port 9000, then check if test_callback is called.

**Comment by john0312** [comment]
Note: If you need HttpListener to be merged soon because other classes depends on it, then you can separate the unit test into a separate PR so that you can get HttpListener merged sooner.

File: src/Implement/HttpListener.cpp
Comment (john0312) [reply src/Implement/HttpListener.cpp]: Test that the callback is triggered when a connection comes in.

File: src/Implement/UnitTestHttpListener.cpp
Comment (john0312) [reply src/Implement/UnitTestHttpListener.cpp:52]: Please implement the unit test as discussed in class.

File: src/Implement/HttpListener.cpp
Comment (younger15) [reply src/Implement/HttpListener.cpp]: Done

File: src/Implement/UnitTestHttpListener.cpp
Comment (younger15) [reply src/Implement/UnitTestHttpListener.cpp:52]: Done

File: src/Implement/UnitTestHttpListener.cpp
Comment (john0312) [inline src/Implement/UnitTestHttpListener.cpp]: This can go in an anonymous namespace.

File: src/Implement/UnitTestHttpListener.cpp
Comment (john0312) [inline src/Implement/UnitTestHttpListener.cpp]: Consider pulling this out as a global anonymous constant so that you will not change one and not the other.
```
namespace {
// Test TCP Port for testing the listener.
constexpr uint16_t kTestPort = 9000;
}
```

File: src/Implement/UnitTestHttpListener.cpp
Comment (john0312) [inline src/Implement/UnitTestHttpListener.cpp:22]: This can go in an anonymous namespace.

File: src/Implement/UnitTestHttpListener.cpp
Comment (john0312) [inline src/Implement/UnitTestHttpListener.cpp]: This can go in an anonymous namespace.

**Review by Jazz86 — APPROVED** [review]
Approved. Assuming the listed problem(printf, namespace) has been fixed.

File: src/Implement/UnitTestHttpListener.cpp
Comment (younger15) [reply src/Implement/UnitTestHttpListener.cpp]: Done

File: src/Implement/UnitTestHttpListener.cpp
Comment (younger15) [reply src/Implement/UnitTestHttpListener.cpp]: Done

File: src/Implement/UnitTestHttpListener.cpp
Comment (younger15) [reply src/Implement/UnitTestHttpListener.cpp:22]: Done

File: src/Implement/UnitTestHttpListener.cpp
Comment (younger15) [reply src/Implement/UnitTestHttpListener.cpp]: Done

File: src/Implement/UnitTestHttpListener.cpp
Comment (john0312) [reply src/Implement/UnitTestHttpListener.cpp]: This is still outside of the anonymous namespace.

Comment (younger15) [reply src/Implement/UnitTestHttpListener.cpp]: Done

**Review by john0312 — APPROVED** [review]
Approve assuming you clang format before merging.

**Pull request merged**
